platform-tests/src/test/java/org/junit/platform/engine/support/hierarchical/ForkJoinPoolHierarchicalTestExecutorServiceTests.java

```diff
@@ -67,44 +108,128 @@ void defersTasksWithIncompatibleLocks() throws Exception {
 			deferred.countDown();
 		};
 
-		var isolatedTask = new DummyTestTask("isolatedTask", globalReadWriteLock,
-			t -> threadNamesByTaskIdentifier.put(t.identifier(), Thread.currentThread().getName()));
+		var incompatibleTask = new DummyTestTask("incompatibleTask", incompatibleLock);
+
+		var tasks = runWithAttemptedWorkStealing(taskEventListener, incompatibleTask, initialLock, () -> {
+			try {
+				deferred.await();
+			}
+			catch (InterruptedException e) {
+				System.out.println("Interrupted while waiting for task to be deferred");
+			}
+		});
+
+		assertEquals(incompatibleTask, deferredTask.get());
+		assertEquals(tasks.get("nestedTask").threadName, tasks.get("leafTask2").threadName);
+		assertNotEquals(tasks.get("leafTask1").threadName, tasks.get("leafTask2").threadName);
+	}
+
+	static List<Arguments> compatibleLockCombinations() {
+		return List.of(//
+			arguments(//
+				Set.of(GLOBAL_READ), //
+				Set.of(new ExclusiveResource("a", LockMode.READ)) //
+			), //
+			arguments(//
+				Set.of(GLOBAL_READ), //
+				Set.of(GLOBAL_READ, new ExclusiveResource("a", LockMode.READ)) //
+			), //
+			arguments(//
+				Set.of(GLOBAL_READ, new ExclusiveResource("a", LockMode.READ)), //
+				Set.of(GLOBAL_READ, new ExclusiveResource("a", LockMode.READ)) //
+			), //
+			arguments(//
+				Set.of(GLOBAL_READ, new ExclusiveResource("a", LockMode.READ)), //
+				Set.of(GLOBAL_READ, new ExclusiveResource("b", LockMode.READ)) //
+			)//
+		);
+	}
+
+	@ParameterizedTest
+	@MethodSource("compatibleLockCombinations")
+	void canWorkStealTaskWithCompatibleLocks(Set<ExclusiveResource> initialResources,
+			Set<ExclusiveResource> compatibleResources) throws Exception {
+
+		var lockManager = new LockManager();
+		var initialLock = lockManager.getLockForResources(initialResources);
+		var compatibleLock = lockManager.getLockForResources(compatibleResources);
+
+		var deferredTask = new AtomicReference<TestTask>();
+
+		var workStolen = new CountDownLatch(1);
+		var compatibleTask = new DummyTestTask("compatibleTask", compatibleLock, workStolen::countDown);
+
+		var tasks = runWithAttemptedWorkStealing(deferredTask::set, compatibleTask, initialLock, () -> {
+			try {
+				workStolen.await();
+			}
+			catch (InterruptedException e) {
+				System.out.println("Interrupted while waiting for work to be stolen");
+			}
+		});
+
+		assertNull(deferredTask.get());
+		assertEquals(tasks.get("nestedTask").threadName, tasks.get("leafTask2").threadName);
+		assertNotEquals(tasks.get("leafTask1").threadName, tasks.get("leafTask2").threadName);
+	}
+
+	private static Map<String, DummyTestTask> runWithAttemptedWorkStealing(TaskEventListener taskEventListener,
```

Comment:
Right now we only try one level of nesting, but shouldn't we also check multiple levels?

Reply:
You mean with additional locks on every level?

Reply:
Ideally, we'd test this; the parentheses mean that they are deferred at that level.

A -> (B) -> C -> (D)

Can we achieve it in an easier scenario?

Reply:
Done in 7c654e450e7ff5dd4ebd950498895f190fcaf455.